src/mesh-formats.md

```diff
@@ -169,58 +169,169 @@ we get:
+`vertex_parents` section would look like this:
+
+```sh
+vertex_parents
+4
+6 0 3
+7 1 4
+8 0 1
+9 6 7
+```
+
+The hanging vertices can appear in any order in this section. The only
+limitation is that the first N vertex indices (not listed in this section) be
+reserved for the top-level vertices (those with no parents, typically the
+vertices of the coarse mesh).
+
+The next section is optional and can be safely omitted when creating the mesh
+file manually. The `root_state` section affects leaf ordering when traversing
+the refinement trees and is used to optimize the SFC-based partitioning. There
+is one number per root element. The default state for all root elements is zero.
+
+```sh
+root_state
+<number of records>
+<root element state>
 ...
 ```
 
-These are followed by the standard sections `vertices` and `nodes` of the format
-MFEM mesh v1.0.
-
-The new section `vertex_parents` identifies all vertices (by their zero-based
-index) that have been created as new mid-edge vertices by adaptive refinement
-of elements. Each such vertex has exactly two "parents" identified again by two
-zero-based indices. This information is needed to construct constraining
-relations in a mesh with hanging nodes. The order the vertex-parent relations
-are stated in the file is not significant.
-
-The second optional section `coarse_elements` describes the element refinement
-hierarchy. While the standard section `elements` lists all leaf elements of the
-refinement tree, this section describes all elements that have been refined and
-are no longer active. Each line describes one such virtual element, its
-refinement type and up to 8 children. Child indices between 0 and _N-1_ refer
-to the _N_ active elements in the `elements` section. A coarse element has an
-implied index starting with _N_. A coarse element can refer to another coarse
-element of _index >= N_, but only after such child has been defined in the
-`coarse_elements` section. The hierarchy is thus represented from the bottom
-up. The refinement types are: 1=X, 2=Y, 4=Z, 3=XY, 5=XZ, 6=YZ, 7=XYZ, where X,
-Y, Z refer to one or more splits in the respective axes of the element
-reference domain. If the entire section is missing, MFEM will not be able to
-derefine the mesh.
+Finally, we have the `coordinates` section which assigns physical positions
+to the N top-level vertices. Note that the positions of hanging vertices are
+always derived from their parent vertices and are not listed in the mesh file.
 
-![](img/fichera-amr.png)
+```sh
+coordinates
+<number of top-level vertices>
+<vdim>
+<coordinate 1> ... <coordinate <vdim>>
+...
+```
+
+If the mesh is curvilinear, the `coordinates` section can be replaced with
+an alternative section called `nodes`. The `nodes` keyword is then followed
+by a serialized `GridFunction` representing a vector-valued finite element
+function defining the curvature of the elements, similarly as in the conforming
+case.
 
```

Comment:
Perhaps just a brief comment about `mfem_mesh_end` since this also differs from the "MFEM mesh v1.0" standard.

Reply:
Forgot that too, thanks!